lib/Dialect/Openfhe/Transforms/ConfigureCryptoContext.cpp

```diff
@@ -38,232 +38,301 @@ namespace openfhe {
-    keySwitchCount = openfheParamsAttr.getKeySwitchCount();
-    // remove the attribute after reading
-    op->removeAttr(mgmt::MgmtDialect::kArgOpenfheParamsAttrName);
+  // Helper function to find all the rotation indices in the function
+  SmallVector<int64_t> findAllRotIndices(func::FuncOp op) {
+    std::set<int64_t> distinctRotIndices;
+    op.walk([&](openfhe::RotOp rotOp) {
+      distinctRotIndices.insert(rotOp.getIndex().getInt());
+      return WalkResult::advance();
+    });
+    SmallVector<int64_t> rotIndicesResult(distinctRotIndices.begin(),
+                                          distinctRotIndices.end());
+    return rotIndicesResult;
   }
 
-  Type openfheParamsType = openfhe::CCParamsType::get(builder.getContext());
-  Value ccParams = builder.create<openfhe::GenParamsOp>(
-      openfheParamsType, mulDepth, plainMod, insecure, evalAddCount,
-      keySwitchCount, encryptionTechniqueExtended);
-  Value cryptoContext = builder.create<openfhe::GenContextOp>(
-      openfheContextType, ccParams,
-      BoolAttr::get(builder.getContext(), hasBootstrapOp));
-
-  builder.create<func::ReturnOp>(cryptoContext);
-  return success();
-}
-
-// function that configures the crypto context with proper keygeneration
-LogicalResult generateConfigFunc(
-    func::FuncOp op, const std::string &configFuncName, bool hasRelinOp,
-    SmallVector<int64_t> rotIndices, bool hasBootstrapOp, int levelBudgetEncode,
-    int levelBudgetDecode, ImplicitLocOpBuilder &builder) {
-  Type openfheContextType =
-      openfhe::CryptoContextType::get(builder.getContext());
-  Type privateKeyType = openfhe::PrivateKeyType::get(builder.getContext());
-
-  SmallVector<Type> funcArgTypes;
-  funcArgTypes.push_back(openfheContextType);
-  funcArgTypes.push_back(privateKeyType);
-
-  SmallVector<Type> funcResultTypes;
-  funcResultTypes.push_back(openfheContextType);
-
-  FunctionType configFuncType =
-      FunctionType::get(builder.getContext(), funcArgTypes, funcResultTypes);
-  auto configFuncOp =
-      builder.create<func::FuncOp>(configFuncName, configFuncType);
-  builder.setInsertionPointToEnd(configFuncOp.addEntryBlock());
-
-  Value cryptoContext = configFuncOp.getArgument(0);
-  Value privateKey = configFuncOp.getArgument(1);
-
-  if (hasRelinOp || hasBootstrapOp) {
-    builder.create<openfhe::GenMulKeyOp>(cryptoContext, privateKey);
+  // Helper function to check if the function has BootstrapOp
+  bool hasBootstrapOp(func::FuncOp op) {
+    bool result = false;
+    op.walk<WalkOrder::PreOrder>([&](Operation *op) {
+      if (isa<openfhe::BootstrapOp>(op)) {
+        result = true;
+        return WalkResult::interrupt();
+      }
+      return WalkResult::advance();
+    });
+    return result;
   }
-  if (!rotIndices.empty()) {
-    builder.create<openfhe::GenRotKeyOp>(cryptoContext, privateKey, rotIndices);
+
+  // global pass object to store the configuration, for other functions
+  // to access easily. Good design?
```

Comment:
I think it is fine since this is a member variable of the pass instance, not a global variable or a static variable of the class. So each instance of this pass will get a different config object (instantiated with default-constructed values for all fields). You could make this slightly clearer by defining the struct outside this class, and then just having a private member variable declared under the class constructor.

Reply:
Moved that outside the class and have a `Config config` in private member field